### ISSUE TITLE
Weighted CopyGenerator Loss

### ISSUE THREAD

**Issue by oaishi** [OpenNMT/OpenNMT-py#1842]
We are using OpenNMT copy generator loss as our loss criterion [[onmt/utils/loss.py#L34]](https://github.com/OpenNMT/OpenNMT-py/blob/592a2f62f098c69beeb5487f93d6c74321363630/onmt/utils/loss.py#L34).  Currently, we need to incorporate weighted loss per token in copy generator loss. We looked over some related issues regarding weighted nllloss function in OpenNMT [[#271 ]](https://github.com/OpenNMT/OpenNMT-py/issues/271) and [[#290 ]](https://github.com/OpenNMT/OpenNMT-py/pull/290). However, CopyGeneratorLoss is much complex than that, and we were confused about how we can apply weighted loss in this case without hampering other OpenNMT functionalities. We will be grateful if you kindly help us on how we can tackle our issues and pass the token weights successfully.

**Comment by pltrdy** [comment]
I'm not sure to understand it right. 
You want to use a weight `w_i` for each example `i`, i.e. each pair `(src, tgt)`, `weighted_loss = loss * w_i` ?

In this case, it would require a product between loss and weights at https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/copy_generator.py#L249. At this point the loss is a vector of size `batch_size` so we can weigh it with a vector `W = [w_i, ..., w_{i+batch_size}]`. The rest of the function then sum individual example losses into a single batch-loss.

**Comment by oaishi** [comment]
Thank you so much for your fast and helpful reply!

I will try to describe the situation a bit more clearly for better understanding. We have a list of n tokens -  t_0, t_1, ...., t_n and weights associated with them, w_0, w_1, ...., w_n. If the target token is t_n and predicted token is o_n, we want to calculate the loss as = w_n * loss ( t_n, o_n). 

The model configuration we are using -

```
!onmt_train -data data/cc/pretrain_data \
            -save_model models/cc/pretrained_model_benchmark \
            -batch_size 60 \
            -report_every 200 \
            -world_size 1 \
            -gpu_ranks 0 \
            -copy_attn -global_attention mlp \
            -word_vec_size 128 \
            -rnn_size 512 \
            -layers 1 \
            -encoder_type brnn \
            -train_steps 500000 \
            -max_grad_norm 2 \
            -dropout 0. \
            -optim adagrad \
            -learning_rate 0.15 \
            -adagrad_accumulator_init 0.1 \
            -reuse_copy_attn \
            -copy_loss_by_seqlength\
            -bridge \
            -seed 777 \
```
We want to use this weighted loss because the problem we are facing is - a greater portion of our dataset (about one-third) contains the same and single output token (but other data output lengths are around 50-80 tokens) which is causing the results to bias towards that token. So, we are facing two problems at the same time - biased token and data length imbalance problem. 

As a workaround, we were thinking of using weighted loss to minimize that tokens affect.  Is there any other way to mitigate the data length imbalance problem too?

**Comment by pltrdy** [comment]
Ok got it. 

Then your weights, as you explained it (different from my first understanding), would make sense for your problem. However, it would need to add some code at the preprocessing step to load a vocab like file i.e. some kind of `t_i w_i` list for each `t_i` of your target vocabulary, and to associate this list with the target vocab object.

Then, when calculating loss, you must get target weights similarly to target scores (i.e. https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/copy_generator.py#L156) then your final scores are just `scores = weights * scores` (tensors must have the same size).

**Comment by oaishi** [comment]
Thank you so much for your detailed reply! It will help us a lot.
